### PR TITLE
[Downgrade PHP7.3] Remove unneeded params from `list()`, and even remove `list()`

### PR DESCRIPTION
**Pull request by leoloso**
Further enhancement. After downgrading `list()` reference assignment (#4371) can remove the right-side-most params by reference from `list()`:

```diff
class SomeClass
{
    public function run($string)
    {
        $array = [1, 2, 3];
-        list($a, &$b) = $array;
+        list($a) = $array;
+        $b =& $array[1];
    }
}
```

If all params are by reference, can directly remove `list()`:

```diff
class SomeClass
{
    public function run($string)
    {
        $array = [1, 2, 3];
-        list(&$a, &$b) = $array;
+        $a =& $array[0];
+        $b =& $array[1];
    }
}
```

If there are nested `list()`s, if they have only references, they can be removed:

```diff
class SomeClass
{
    public function run($string)
    {
        $array = [1, 2, 3, 4];
-        list($a, &$b, list(&$c, &$d)) = $array;
+        list($a) = $array;
+        $b =& $array[1];
+        $c =& $array[2];
+        $d =& $array[3];
    }
}
```

Otherwise can't:

```diff
class SomeClass
{
    public function run($string)
    {
        $array = [1, 2, 3, 4];
-        list($a, &$b, list($c, &$d)) = $array;
+        list($a, $b, list($c, $d)) = $array;
+        $b =& $array[1];
+        $d =& $array[3];
    }
}
```

### CODE REVIEW

File: rules/downgrade-php73/src/Rector/List_/DowngradeListReferenceAssignmentRector.php [hunk 1]
```diff
@@ -217,4 +273,40 @@ private function createAssignRefWithArrayDimFetch(
...
+            $hasItemByRef = false;
+            if ($condition === self::ALL) {
+                $hasItemByRef = $this->hasAllItemsByRef($nestedList->items);
+            } elseif ($condition === self::ANY) {
+                $hasItemByRef = $this->hasAnyItemByRef($nestedList->items);
+            }
+            return $hasItemByRef ? $arrayItem : null;
+        }
+        return $arrayItem->value instanceof Variable && $arrayItem->byRef ? $arrayItem : null;
+    }
```
Comment (TomasVotruba) [inline rules/downgrade-php73/src/Rector/List_/DowngradeListReferenceAssignmentRector.php:311]: Btw, this starts to slow like a God class. I need to lower class complexity for PHPStan.

Could you decouple some logic you find most valuable to standalone services?

**Comment by leoloso** [comment]
Done, please check again.

I had also planned to remove params by reference always, having an empty param instead (`,,`) like this:

```diff
class SomeClass
{
    public function run($string)
    {
        $array = [1, 2, 3];
-        list($a, &$b, $c) = $array;
+        list($a,, $c) = $array;
+        $b =& $array[1];
    }
}
```

But this doesn't work with string keys, the code below throws an error on runtime:

```diff
class SomeClass
{
    public function run($string)
    {
        $array = ['a' => 1, 'b' => 2, 'c' => 3];
-        list('a' => $a, 'b' => &$b, 'c' => $c) = $array;
+        list('a' => $a,, 'c' => $c) = $array;
+        $b =& $array['b'];
    }
}
```

The code was going to get more complex to tell which case it is, and it's not worth it, so I didn't do it then. So no need to implement the `RemoveRightSideMostEmptyParamsFromListRector`, and no need (yet) to implement sets of sets or some solution like that.

Btw, `composer phpstan` is failing but it's unrelated to this PR, please check. Or do I need to integrate the changes from master?

**Comment by TomasVotruba** [comment]
> Btw, composer phpstan is failing but it's unrelated to this PR, please check. Or do I need to integrate the changes from master?

It would be better to pass CI, even if the fail was not caused here.

Can you handle it?

**Comment by leoloso** [comment]
> Can you handle it?

Can try. Are you aware of the issue already, what might it be?

It happens on master also, being up-to-date with the repo. I did `composer update` and then `composer phpstan`, and I already get the error. Same error happened in my Travis build of the PR, so it may be from some updated dependency, from the last few hours.

The error is this one:

```
➜  rector git:(master) composer phpstan
> vendor/bin/phpstan analyse --ansi --error-format symplify
Note: Using configuration file /Users/mac/GitRepos/GitHub/Libraries/rectorphp/rector/phpstan.neon.
 2586/2586 [▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓] 100%

     -
         message: '#Do not use @method tag in class docblock#'
         paths:
             - src/PhpParser/Builder/ClassBuilder.php # 15
             - src/PhpParser/Builder/MethodBuilder.php # 15
             - src/PhpParser/Builder/NamespaceBuilder.php # 15
             - src/PhpParser/Builder/ParamBuilder.php # 15
             - src/PhpParser/Builder/PropertyBuilder.php # 15
             - src/PhpParser/Builder/TraitUseBuilder.php # 15
             - src/PhpParser/Builder/UseBuilder.php # 15

     -
         message: '#Property with protected modifier is not allowed\. Use interface instead#'
         paths:
             - packages/better-php-doc-parser/src/PhpDocNodeFactory/AbstractPhpDocNodeFactory.php # 39
             - packages/better-php-doc-parser/src/PhpDocNodeFactory/AbstractPhpDocNodeFactory.php # 44
             - packages/better-php-doc-parser/src/PhpDocNodeFactory/AbstractPhpDocNodeFactory.php # 49
             - packages/better-php-doc-parser/tests/PhpDocInfo/PhpDocInfoPrinter/AbstractPhpDocInfoPrinterTest.php # 24
             - packages/better-php-doc-parser/tests/PhpDocInfo/PhpDocInfoPrinter/AbstractPhpDocInfoPrinterTest.php # 29
             - packages/file-system-rector/src/Rector/AbstractFileMovingFileSystemRector.php # 15
             - packages/file-system-rector/src/Rector/AbstractFileSystemRector.php # 34
             - packages/file-system-rector/src/Rector/AbstractFileSystemRector.php # 39
             - packages/node-type-resolver/tests/PerNodeTypeResolver/AbstractNodeTypeResolverTest.php # 19
             - packages/post-rector/src/Rector/AbstractRector/NodeCommandersTrait.php # 36
             - packages/post-rector/src/Rector/AbstractRector/NodeCommandersTrait.php # 36
             - packages/reporting/src/Rector/AbstractRector/NodeReportCollectorTrait.php # 19
             - packages/reporting/src/Rector/AbstractRector/NodeReportCollectorTrait.php # 19
             - packages/vendor-locker/src/NodeVendorLocker/AbstractNodeVendorLockResolver.php # 21
             - packages/vendor-locker/src/NodeVendorLocker/AbstractNodeVendorLockResolver.php # 26
             - packages/vendor-locker/src/NodeVendorLocker/AbstractNodeVendorLockResolver.php # 31
             - rules/defluent/src/Rector/AbstractFluentChainMethodCallRector.php # 27
             - rules/defluent/src/Rector/AbstractFluentChainMethodCallRector.php # 32
             - rules/defluent/src/Rector/AbstractFluentChainMethodCallRector.php # 37
             - rules/defluent/src/Rector/AbstractFluentChainMethodCallRector.php # 42
             - rules/defluent/src/Rector/AbstractFluentChainMethodCallRector.php # 47
             - rules/downgrade-php71/src/Rector/FunctionLike/AbstractMaybeAddDocBlockRector.php # 20
             - rules/naming/src/PropertyRenamer/AbstractPropertyRenamer.php # 31
             - rules/naming/src/PropertyRenamer/AbstractPropertyRenamer.php # 36
             - rules/nette-code-quality/src/Rector/ArrayDimFetch/AbstractArrayDimFetchToAnnotatedControlVariableRector.php # 32
             - rules/nette-code-quality/src/Rector/ArrayDimFetch/AbstractArrayDimFetchToAnnotatedControlVariableRector.php # 37
             - rules/nette-code-quality/src/Rector/ArrayDimFetch/AbstractArrayDimFetchToAnnotatedControlVariableRector.php # 42
             - rules/order/src/Rector/AbstractConstantPropertyMethodOrderRector.php # 16
             - rules/order/src/Rector/AbstractConstantPropertyMethodOrderRector.php # 21
             - rules/php80/src/MatchAndRefactor/StrStartsWithMatchAndRefactor/AbstractMatchAndRefactor.php # 22
             - rules/php80/src/MatchAndRefactor/StrStartsWithMatchAndRefactor/AbstractMatchAndRefactor.php # 27
             - rules/php80/src/MatchAndRefactor/StrStartsWithMatchAndRefactor/AbstractMatchAndRefactor.php # 32
             - rules/symfony/src/Rector/MethodCall/AbstractFormAddRector.php # 23
             - rules/symfony/src/Rector/MethodCall/AbstractToConstructorInjectionRector.php # 27
             - rules/type-declaration/src/AlreadyAssignDetector/AbstractAssignDetector.php # 18
             - rules/type-declaration/src/Rector/FunctionLike/AbstractTypeDeclarationRector.php # 24
             - rules/type-declaration/src/Rector/FunctionLike/AbstractTypeDeclarationRector.php # 29
             - rules/type-declaration/src/TypeInferer/AbstractTypeInferer.php # 18
             - rules/type-declaration/src/TypeInferer/AbstractTypeInferer.php # 23
             - rules/type-declaration/src/TypeInferer/AbstractTypeInferer.php # 28
             - rules/type-declaration/src/TypeInferer/AbstractTypeInferer.php # 33
             - rules/type-declaration/src/TypeInferer/AbstractTypeInferer.php # 38
             - src/Console/Command/AbstractCommand.php # 22
             - src/Rector/AbstractRector.php # 65
             - src/Rector/AbstractRector.php # 70
             - src/Rector/AbstractRector.php # 75
             - src/Rector/AbstractRector.php # 80
             - src/Rector/AbstractRector.php # 85
             - src/Rector/AbstractRector/BetterStandardPrinterTrait.php # 22
             - src/Rector/AbstractRector/BetterStandardPrinterTrait.php # 27
             - src/Rector/AbstractRector/BetterStandardPrinterTrait.php # 32
             - src/Rector/AbstractRector/BetterStandardPrinterTrait.php # 22
             - src/Rector/AbstractRector/BetterStandardPrinterTrait.php # 27
             - src/Rector/AbstractRector/BetterStandardPrinterTrait.php # 32
             - src/Rector/AbstractRector/ComplexRemovalTrait.php # 39
             - src/Rector/AbstractRector/ComplexRemovalTrait.php # 44
             - src/Rector/AbstractRector/ComplexRemovalTrait.php # 49
             - src/Rector/AbstractRector/ComplexRemovalTrait.php # 39
             - src/Rector/AbstractRector/ComplexRemovalTrait.php # 44
             - src/Rector/AbstractRector/ComplexRemovalTrait.php # 49
             - src/Rector/AbstractRector/NodeCollectorTrait.php # 18
             - src/Rector/AbstractRector/NodeCollectorTrait.php # 18
             - src/Rector/AbstractRector/NodeFactoryTrait.php # 33
             - src/Rector/AbstractRector/NodeFactoryTrait.php # 33
             - src/Rector/AbstractRector/PhpDocTrait.php # 24
             - src/Rector/AbstractRector/PhpDocTrait.php # 29
             - src/Rector/AbstractRector/PhpDocTrait.php # 34
             - src/Rector/AbstractRector/PhpDocTrait.php # 24
             - src/Rector/AbstractRector/PhpDocTrait.php # 29
             - src/Rector/AbstractRector/PhpDocTrait.php # 34
             - src/Testing/PHPUnit/AbstractGenericRectorTestCase.php # 38
             - src/Testing/PHPUnit/AbstractGenericRectorTestCase.php # 43
             - src/Testing/PHPUnit/AbstractGenericRectorTestCase.php # 48
             - src/Testing/PHPUnit/AbstractGenericRectorTestCase.php # 53
             - src/Testing/PHPUnit/AbstractGenericRectorTestCase.php # 58
             - src/Testing/PHPUnit/AbstractGenericRectorTestCase.php # 63
             - src/Testing/PHPUnit/AbstractGenericRectorTestCase.php # 68
             - src/Testing/PHPUnit/AbstractNodeVisitorTestCase.php # 25
             - src/Testing/PHPUnit/AbstractNodeVisitorTestCase.php # 30
             - src/Testing/PHPUnit/AbstractRectorTestCase.php # 24
             - utils/doctrine-annotation-parser-syncer/src/FileSyncer/AbstractClassSyncer.php # 19


                                                                                
 [ERROR] Found 89 errors                                                        
                                                                                

                                                                                
 [WARNING] Internal error: Internal error: Call to a member function toString() 
           on null in file                                                      
           /Users/mac/GitRepos/GitHub/Libraries/rectorphp/rector/src/PhpParser/N
           odeTraverser/CallableNodeTraverser.php                               
           Run PHPStan with --debug option and post the stack trace to:         
           https://github.com/phpstan/phpstan/issues/new                        
                                                                                

Script vendor/bin/phpstan analyse --ansi --error-format symplify handling the phpstan event returned with error code 1
```

**Comment by TomasVotruba** [comment]
I haven't check it yet, on mobile.

That's a lot of new errors, that could be handled in own PR., better ignore them now to merge this soon.

Well handle it new PR.

**Comment by TomasVotruba** [comment]
Sure

> better ignore them now to merge this soon.

**Comment by TomasVotruba** [comment]
Thank you

**Pull request merged**